fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/MutableRecordStoreState.java

```diff
@@ -42,13 +42,17 @@
 
     private final AtomicLong users = new AtomicLong();
 
+    public MutableRecordStoreState(@Nullable RecordMetaDataProto.DataStoreInfo storeHeader, @Nullable Map<String, IndexState> indexStateMap, @Nullable Map<String, IndexMetaDataProto.IndexMetaData> indexMetaDataMap) {
```

Comment:
This is just a thought--and something we discussed--but in a number of places we are carrying around these two discrete pieces of information.  In this case,  map of name->state and name->metadata.  Perhaps we should, instead, only carry around the metadata -- after all, the metadata already contains the state itself.   In this way, they are always kept together and, presumably, in sync.  That is, it would become harder to have the wrong metadata associated with the current state.

Note that I am not suggesting changing the on-disk representation, rather how we convey this information around in the code.  I am also not suggesting that the metadata always needs to be read when opening the store.  We could, instead, have the in-memory object carry around whether or not the metadata has been loaded and either lazily load it or give an error if you try to access information that hasn't been loaded yet.

Anyway, I'm not 100% sure that this can be easily done, but worth thinking / talking about.